internal/datastore/common/sql.go

```diff
@@ -131,57 +133,79 @@ func (sqf SchemaQueryFilterer) TupleOrder(order options.SortOrder) SchemaQueryFi
 }
 
 func (sqf SchemaQueryFilterer) After(cursor *core.RelationTuple, order options.SortOrder) SchemaQueryFilterer {
+	columnsAndValues := []struct {
+		name  string
+		value string
+	}{
+		{
+			sqf.schema.colNamespace, cursor.ResourceAndRelation.Namespace,
+		},
+		{
+			sqf.schema.colObjectID, cursor.ResourceAndRelation.ObjectId,
+		},
+		{
+			sqf.schema.colRelation, cursor.ResourceAndRelation.Relation,
+		},
+		{
+			sqf.schema.colUsersetNamespace, cursor.Subject.Namespace,
+		},
+		{
+			sqf.schema.colUsersetObjectID, cursor.Subject.ObjectId,
+		},
+		{
+			sqf.schema.colUsersetRelation, cursor.Subject.Relation,
+		},
+	}
+
 	switch sqf.schema.paginationFilterType {
 	case TupleComparison:
 		switch order {
 		case options.ByResource:
-			comparisonTuple := fmt.Sprintf(
-				"(%s, %s, %s, %s, %s) > (?, ?, ?, ?, ?)",
-				sqf.schema.colObjectID,
-				sqf.schema.colRelation,
-				sqf.schema.colUsersetNamespace,
-				sqf.schema.colUsersetObjectID,
-				sqf.schema.colUsersetRelation,
-			)
-			sqf.queryBuilder = sqf.queryBuilder.Where(
-				comparisonTuple,
-				cursor.ResourceAndRelation.ObjectId,
-				cursor.ResourceAndRelation.Relation,
-				cursor.Subject.Namespace,
-				cursor.Subject.ObjectId,
-				cursor.Subject.Relation,
-			)
+			// For performance reasons, remove any column names that have static values in the query.
+			columnNames := make([]string, 0, len(columnsAndValues))
+			comparisonSlots := make([]string, 0, len(columnsAndValues))
+			valueSlots := make([]any, 0, len(columnsAndValues))
+
+			for _, cav := range columnsAndValues {
+				if sqf.filteringColumnCounts[cav.name] != 1 {
+					columnNames = append(columnNames, cav.name)
+					comparisonSlots = append(comparisonSlots, "?")
+					valueSlots = append(valueSlots, cav.value)
+				}
+			}
+
+			if len(comparisonSlots) > 0 {
+				comparisonTuple := "(" + strings.Join(columnNames, ",") + " > " + strings.Join(comparisonSlots, ",") + ")"
```

Comment:
Maybe this is micro-optimizations, but believe at large-scale all those allocations do matter. We can spare one slice allocation using `strings.Repeat`. It's also just a bit faster.

```go
func BenchmarkRepeat(b *testing.B) {
	columnNames := make([]string, 6, 6)
	for i := 0; i < b.N; i++ {
		_ = strings.Repeat(",?", len(columnNames))[1:]
	}
}

func BenchmarkJoin(b *testing.B) {
	columnNames := []string{"?", "?", "?", "?", "?", "?"}
	for i := 0; i < b.N; i++ {
		strings.Join(columnNames, ",")
	}
}
```
```
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkRepeat
BenchmarkRepeat-16    	23180281	        50.46 ns/op
PASS

cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkJoin
BenchmarkJoin-16    	18482060	        61.02 ns/op
PASS
```